rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,554 @@
+performing http traffic balancing between the workers for us.
+
+![image](../images/15_clustering/cluster-mode.png)
+
+The coordinator sole responsibility is to orchestrate the workers. It would not be a 'super' worker handling both 
+the job of a worker while being in charge of managing the other workers.
+
+# Performance testing
+
+## Local testing
+
+These tests were performed against a local development machine, with a 8-core cpu (2.4 GHz 8-Core Intel Core i9 - 32 GB 2400 MHz DDR4),
+using the default configuration of the `kibana-load-testing` tool.
+
+### Raw results
+
+#### Non-clustered mode
+
+![image](../images/15_clustering/perf-no-clustering.png)
+
+#### Clustered mode, 2 workers
+
+![image](../images/15_clustering/perf-clustering-2-worker.png)
+
+#### Clustered mode, 4 workers
+
+![image](../images/15_clustering/perf-4-workers.png)
+
+### Analysis 
+
+- Between non-clustered and 2-worker cluster mode, we observe a 20/25% gain in the 50th percentile response time. 
+  Gain for the 75th and 95th are between 10% and 40%
+- Between 2-worker and 4-workers cluster mode, the gain on 50th is negligible, but the 75th and the 95th are 
+  significantly better on the 4-workers results, sometimes up to 100% gain (factor 2 ratio)
+
+Overall, switching to 2 workers comes with the most significant improvement in the 50th pct, 
+and increasing further to 4 workers decreases even more significantly the highest percentiles. 
+Even if increasing the number of workers doesn’t just linearly increase the performances 
+(which totally make sense, most of our requests response time is caused by awaiting ES response), 
+the improvements of the clustering mode on performance under heavy load are far from negligible.
+
+## Testing against cloud
+
+There is currently no easy way to test the performance improvements this could provide on Cloud, as we can't
+deploy custom builds or branches on Cloud at the moment.
+
+On Cloud, Kibana is running in a containerised environment using CPU CFS quota and CPU shares. 
+
+If we want to investigate the potential perf improvement on Cloud further, our only option would be to setup a 
+similar-ish environment locally (which wasn't done during the initial investigation)
+
+# Implementation proposal
+
+## enabling clustering mode
+
+Enabling clustering mode will be done using the `clustering.enabled` configuration property.
+
+The whole `clustering` configuration would look like:
+
+```ts
+export const config = {
+  path: 'clustering',
+  schema: schema.object({
+    enabled: schema.boolean({ defaultValue: false }),
+    workers: schema.conditional(
+      schema.siblingRef('enabled'),
+      false,
+      schema.number({ defaultValue: 1 }),
+      schema.number()
+    )
+  }),
+};
+```
+
+Notes:
+- As there isn't really a good way to automatically choose the correct number of workers automatically, It will be required 
```

Comment:
Is it possible to make this default to the number of CPUs available?

Reply:
It would be, but from previous discussions we agreed that this could be a potentially dangerous behavior, which is why we decided to force the user to manually specify the value:
- There is the coordinator process to take into account, so the 'ideal' value would be numCpu-1
- We're not sure what else is running on the machine. Having a process per CPU per default could be a potentially dangerous default

Reply:
Updated to reflect a strategy of using all available CPUs (- 1 for the coordinator)